File: javascript/sm3/sm3.js
import {
  stringToBytes,
  bytesToHex,
  bytesToIntBE,
  intToBytesBE,
} from '../common/utils.js'

/** SM3 初始向量 */
const IV = new Uint32Array([
  0x7380166f, 0x4914b2b9, 0x172442d7, 0xda8a0600, 0xa96f30bc, 0x163138aa,
  0xe38dee4d, 0xb0fb0e4e,
])

/** 32 位循环左移 */
function rotl(x, n) {
  return ((x << n) | (x >>> (32 - n))) >>> 0
}

/** 置换函数 P0 */
function P0(x) {
  return x ^ rotl(x, 9) ^ rotl(x, 17)
}

/** 置换函数 P1 */
function P1(x) {
  return x ^ rotl(x, 15) ^ rotl(x, 23)
}

/** 布尔函数 FF (j >= 16) */
function FF1(x, y, z) {
  return (x & y) | (x & z) | (y & z)
}

/** 布尔函数 GG (j >= 16) */
function GG1(x, y, z) {
  return (x & y) | (~x & z)
}

/**
 * SM3 哈希算法
 *
 * 用法：
 *   const h = new SM3()
 *   h.update(bytes).update(moreBytes)
 *   const hash = h.digest()       // Uint8Array(32)
 *   const hex  = h.hexDigest()    // string
 *
 * 便捷方法：
 *   SM3.hash('abc')               // Uint8Array(32)
 *   SM3.hashHex('abc')            // 大写 hex string
 */
export class SM3 {
  constructor() {
    this._buf = new Uint8Array(64) // 512-bit 块缓冲
    this._pos = 0 // 缓冲区写入位置
    this._len = 0 // 已处理字节总长度
    this._V = new Uint32Array(IV) // 当前哈希状态
    this._finished = false
  }

  /**
   * 输入数据（支持单个字节或字节数组）
   * @param {number|Uint8Array|number[]} data - 单个字节 (0-255) 或字节数组
   * @returns {SM3}
   */
  update(data) {
    if (typeof data === 'number') {
      this._buf[this._pos++] = data & 0xff
      this._len++
      if (this._pos === 64) {
        this._compress(this._buf)
        this._pos = 0
      }
      return this
    }
    for (let i = 0; i < data.length; i++) {
      this._buf[this._pos++] = data[i] & 0xff
      this._len++
      if (this._pos === 64) {
        this._compress(this._buf)
        this._pos = 0
      }
    }
    return this
  }

  /**
   * 输入字符串（UTF-8 编码）
   * @param {string} text
   * @returns {SM3}
   */
  updateString(text) {
    return this.update(stringToBytes(text))
  }

  /**
   * 输入字节数组的指定区间（兼容旧 API）
   * @param {Uint8Array|number[]} data
   * @param {number} offset
   * @param {number} length
   * @returns {SM3}
   */
  updateBytes(data, offset, length) {
    for (let i = 0; i < length; i++) {
      this._buf[this._pos++] = data[offset + i] & 0xff
      this._len++
      if (this._pos === 64) {
        this._compress(this._buf)
        this._pos = 0
      }
    }
    return this
  }

  /**
   * 完成哈希计算，返回 32 字节摘要
   * @returns {Uint8Array}
   */
  digest() {
    if (!this._finished) {
      this._pad()
      this._finished = true
    }
    const out = new Uint8Array(32)
    for (let i = 0; i < 8; i++) {
      intToBytesBE(this._V[i], out, i * 4)
    }
    return out
  }

  /**
   * 完成哈希计算，返回大写十六进制字符串
   * @returns {string}
   */
  hexDigest() {
    return bytesToHex(this.digest()).toUpperCase()
  }

  // ---- 兼容旧 API ----

  /** @deprecated 使用 digest() / hexDigest() */
  finish() {
    this._hashBytes = this.digest()
    this._hashValue = bytesToHex(this._hashBytes).toUpperCase()
    // 重置状态以支持复用
    this._V = new Uint32Array(IV)
    this._pos = 0
    this._len = 0
    this._finished = false
    return this
  }

  /** @deprecated 使用 hexDigest() */
  getHashCode() {
    return this._hashValue
  }

  /** @deprecated 使用 digest() */
  getHashBytes() {
    return this._hashBytes
  }

  // ---- 内部方法 ----

  /** 消息填充 */
  _pad() {
    const bitLen = this._len * 8

    // 追加 0x80
    this._buf[this._pos++] = 0x80
    // 如果剩余空间不够放 8 字节长度，先填满当前块
    if (this._pos > 56) {
      while (this._pos < 64) this._buf[this._pos++] = 0
      this._compress(this._buf)
      this._pos = 0
    }
    // 填充零
    while (this._pos < 56) this._buf[this._pos++] = 0
    // 追加 64 位大端序消息长度（JS 安全整数范围内用两个 32 位写入）
    intToBytesBE(Math.floor(bitLen / 0x100000000), this._buf, 56)
    intToBytesBE(bitLen >>> 0, this._buf, 60)
    this._compress(this._buf)
  }

  /** 压缩函数：处理一个 64 字节块 */
  _compress(block) {
    // 消息扩展
    const w = new Int32Array(68)
    for (let j = 0; j < 16; j++) {
      w[j] = bytesToIntBE(block, j * 4)
    }
    for (let j = 16; j < 68; j++) {
      w[j] =
        P1(w[j - 16] ^ w[j - 9] ^ rotl(w[j - 3], 15)) ^
        rotl(w[j - 13], 7) ^
        w[j - 6]
    }

    const w2 = new Int32Array(64)
    for (let j = 0; j < 64; j++) {
      w2[j] = w[j] ^ w[j + 4]
    }

    // 压缩
    let A = this._V[0],
      B = this._V[1],
      C = this._V[2],
      D = this._V[3]
    let E = this._V[4],
      F = this._V[5],
      G = this._V[6],
      H = this._V[7]

    for (let j = 0; j < 64; j++) {
      const A12 = rotl(A, 12)
      const Tj = j < 16 ? rotl(0x79cc4519, j) : rotl(0x7a879d8a, j % 32)
      const SS1 = rotl((A12 + E + Tj) | 0, 7)
      const SS2 = SS1 ^ A12

      const TT1 =
        j < 16
          ? ((A ^ B ^ C) + D + SS2 + w2[j]) | 0
          : (FF1(A, B, C) + D + SS2 + w2[j]) | 0
      const TT2 =
        j < 16
          ? ((E ^ F ^ G) + H + SS1 + w[j]) | 0
          : (GG1(E, F, G) + H + SS1 + w[j]) | 0

      D = C
      C = rotl(B, 9)
      B = A
      A = TT1
      H = G
      G = rotl(F, 19)
      F = E
      E = P0(TT2)
    }

    this._V[0] ^= A
    this._V[1] ^= B
    this._V[2] ^= C
    this._V[3] ^= D
    this._V[4] ^= E
    this._V[5] ^= F
    this._V[6] ^= G
    this._V[7] ^= H
  }

  // ---- 静态便捷方法 ----

  /**
   * 一次性计算字符串的 SM3 哈希
   * @param {string} str
   * @returns {Uint8Array}
   */
  static hash(str) {
    return new SM3().updateString(str).digest()
  }

  /**
   * 一次性计算字符串的 SM3 哈希（大写 hex）
   * @param {string} str
   * @returns {string}
   */
  static hashHex(str) {
    return new SM3().updateString(str).hexDigest()
  }
}

/** @deprecated 使用 new SM3() */
export const getSM3 = () => new SM3()

File: java/src/site/aicc/sm3/SM3.java
package site.aicc.sm3;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;

/** SM3 cryptographic hash algorithm. */
public class SM3 {
    private static final int[] IV = new int[] { 0x7380166F, 0x4914B2B9, 0x172442D7, 0xDA8A0600, 0xA96F30BC, 0x163138AA, 0xE38DEE4D, 0xB0FB0E4E };

    private int[] V;

    private byte[] blockBuffer;
    private int bufferOffset;
    private int dataLength;
    private String hexResult;
    private byte[] hashBytes;

    public SM3() {
        blockBuffer = new byte[64];
        this.bufferOffset = 0;
        this.dataLength = 0;
        V = Arrays.copyOfRange(IV, 0, IV.length);
    }

    public SM3 update(byte data) {
        blockBuffer[bufferOffset++] = data;
        dataLength += 8;
        if (bufferOffset == 64) {
            compressBlock(blockBuffer);
            bufferOffset = 0;
        }
        return this;
    }

    public SM3 update(String data) {
        byte[] bytes = data.getBytes(StandardCharsets.UTF_8);
        return this.update(bytes,0,bytes.length);
    }

    public SM3 update(byte[] data, int inOffset, int length) {
        for (int i = 0; i < length; i++) {
            blockBuffer[bufferOffset++] = data[inOffset + i];
            dataLength += 8;
            if (bufferOffset == 64) {
                compressBlock(blockBuffer);
                bufferOffset = 0;
            }
        }
        return this;
    }

    public String getHashCode() {
        return this.hexResult;
    }

    public byte[] getHashBytes() {
        return this.hashBytes;
    }

    public SM3 finish() {
        byte[] end = Arrays.copyOfRange(blockBuffer, 0, bufferOffset);
        int blockLen = end.length * 8;
        byte one = (byte) 128;
        int fillZeroLen = (512 - (blockLen + 65) % 512) - 7;
        int allLen = fillZeroLen + blockLen + 65 + 7;
        int allByteLen = allLen / 8;
        byte[] buff = new byte[allByteLen];
        for (int i = 0; i < allByteLen; i++) {
            if (i < end.length) {
                buff[i] = end[i];
            } else if (i == end.length) {
                buff[i] = one;
            } else if (i > allByteLen - 5) {
                buff[i] = (byte) ((dataLength >> (allByteLen - i - 1) * 8) & 0xFF);
            } else {
                buff[i] = 0;
            }
        }
        for (int i = 0; i < allLen / 512; i++) {
            byte[] block = Arrays.copyOfRange(buff, i * 512 / 8, (i + 1) * 512 / 8);
            compressBlock(block);
        }
        computeHashResult();
        reset();
        return this;
    }

    private void computeHashResult() {
        this.hashBytes = new byte[32];
        int off = 0;
        for (int i = 0; i < V.length; i++) {
            hashBytes[off] = (byte) ((V[i] >>> 24) & 0xff);
            hashBytes[++off] = (byte) ((V[i] >>> 16) & 0xff);
            hashBytes[++off] = (byte) ((V[i] >>> 8) & 0xff);
            hashBytes[++off] = (byte) (V[i] & 0xff);
            off++;
        }
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < hashBytes.length; i++) {
            sb.append(Integer.toString((hashBytes[i] & 0xff) + 0x100, 16).substring(1));
        }
        this.hexResult = sb.toString().toUpperCase();
    }

    private void reset() {
        V = Arrays.copyOfRange(IV, 0, IV.length);
        bufferOffset = 0;
        dataLength = 0;
    }

    private void compressBlock(byte[] block) {
        int[] w = new int[68];
        int offset = 0;
        for (int j = 0; j < 16; j++) {
            int h1 =  (block[offset] & 0xff) << 24;
            int h2 =  (block[++offset] & 0xff) << 16;
            int h3 =  (block[++offset] & 0xff) << 8;
            int h4  = (block[++offset] & 0xff);
            w[j] = (h1 | h2 | h3 | h4);
            offset++;
        }
        for (int j = 16; j < 68; j++) {
            int wj3 = w[j - 3];
            int r15 = ((wj3 << 15) | (wj3 >>> (32 - 15)));
            int wj13 = w[j - 13];
            int r7 = ((wj13 << 7) | (wj13 >>> (32 - 7)));
            w[j] = P1(w[j - 16] ^ w[j - 9] ^ r15) ^ r7 ^ w[j - 6];
        }
        int[] wPrime = new int[64];
        for (int j = 0; j < wPrime.length; j++) {
            wPrime[j] = w[j] ^ w[j + 4];
        }

        int A = V[0];
        int B = V[1];
        int C = V[2];
        int D = V[3];
        int E = V[4];
        int F = V[5];
        int G = V[6];
        int H = V[7];
        for (int j = 0; j < 64; j++) {
            int A12 = ((A << 12) | (A >>> (32 - 12)));
            int roundConstant = j < 16 ? ((0x79CC4519 << j) | (0x79CC4519 >>> (32 - j))) : ((0x7A879D8A << (j % 32)) | (0x7A879D8A >>> (32 - (j % 32))));
            int rotatedSum = A12 + E + roundConstant;
            int SS1 = ((rotatedSum << 7) | (rotatedSum >>> (32 - 7)));
            int SS2 = SS1 ^ A12;
            int TT1 = j < 16 ? ((A ^ B ^ C) + D + SS2 + wPrime[j]) : (FF1(A, B, C) + D + SS2 + wPrime[j]);
            int TT2 = j < 16 ? ((E ^ F ^ G) + H + SS1 + w[j]) : (GG1(E, F, G) + H + SS1 + w[j]);
            D = C;
            C = ((B << 9) | (B >>> (32 - 9)));
            B = A;
            A = TT1;
            H = G;
            G = ((F << 19) | (F >>> (32 - 19)));
            F = E;
            E = P0(TT2);
        }
        V[0] ^= A;
        V[1] ^= B;
        V[2] ^= C;
        V[3] ^= D;
        V[4] ^= E;
        V[5] ^= F;
        V[6] ^= G;
        V[7] ^= H;
    }

    private int FF1(int X, int Y, int Z) {
        return (X & Y) | (X & Z) | (Y & Z);
    }

    private int GG1(int x, int y, int z) {
        return (x & y) | ((~x) & z);
    }

    private int P0(int x) {
        int x9 = (x << 9) | (x >>> (32 - 9));
        int x17 = (x << 17) | (x >>> (32 - 17));
        return x ^ x9 ^ x17;
    }

    private int P1(int x) {
        int x15 = (x << 15) | (x >>> (32 - 15));
        int x23 = (x << 23) | (x >>> (32 - 23));
        return x ^ x15 ^ x23;
    }

}
